File: ml-benchmark/autolifter/single-pass/dropwhile.f
Config ExtraGrammar = "AutoLifter";
Config SampleIntMin = 0;
Config SampleIntMax = 1;

Inductive List = cons {Int, List} | nil Unit;

single_pass = \v: List -> Int.
  let run = fix (
    \f: List -> Compress List. \xs: List.
    match xs with
      nil _ -> xs
    | cons {h, t} -> cons {h, f t}
    end
  ) in \xs: List.
  v (run xs);

inf = 100;

/*User provided programs*/

dropwhile = fix (
  \f: Int -> List -> Int. \i: Int. \xs: List.
  match xs with
    nil _ -> i
  | cons {h, t} -> if > h 0 then i else f (+ 1 i) t
  end
) 0;

main = single_pass dropwhile;

File: ml-benchmark/autolifter/single-pass/dropwhile.ml
type list_ =
  | Cons of int * list_
  | Nil

let single_pass v =
  let rec run xs =
    match xs with
    | Nil -> xs
    | Cons (h, t) -> Cons (h, run t)
  in
  fun xs -> v (run xs)

let inf = 100

let dropwhile =
  let rec f i xs =
    match xs with
    | Nil -> i
    | Cons (h, t) ->
        if h > 0 then i else f (i + 1) t
  in
  fun xs -> f 0 xs

let program = single_pass dropwhile